### PR TITLE
Focus: Add FocusManager

### PR DESCRIPTION
**Pull request by JeanTracker**
FocusManager replaces NuguFocus by extending the following features:

* Supports freely generate and inject focus channel
* Supports easily control for changing state of focus channel
* Supports monitoring of changing state of focus channel
* Remove internally hard-coded focus name and priority

Signed-off-by: JeanTracker <hyojoong.kim.jean@gmail.com>

### CODE REVIEW

**Comment by JeanTracker** [comment]
The replacement of NuguFocus with FocusManager will be done later.

File: include/clientkit/focus_manager_interface.hh [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+ * FocusManager manage channel's configuration and focus can be acquired or released for each channel.
+ *
+ * @{
+ */
+
+#define DIALOG_CHANNEL_NAME "Dialog" /** @def Default Dialog Channel Name */
+#define DIALOG_CHANNEL_PRIORITY 100 /** @def Default Dialog Channel Priority */
+
+#define COMMUNICATIONS_CHANNEL_NAME "Communications" /** @def Default Communication Channel Name */
+#define COMMUNICATIONS_CHANNEL_PRIORITY 150 /** @def Default Communication Channel Priority */
```
Comment (webispy) [inline include/clientkit/focus_manager_interface.hh:40]: Is there a reason you set COMMUNICATIONS channel priority to 150? Other priority values ​​went up by 100.
e.g. 100(Dialog), 200(Alerts), 300(Content)

File: src/core/focus_manager.cc [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    }
+
+    if (channels_configuration_map.find(channel_name) == channels_configuration_map.end()) {
+        nugu_error("The channel[%s] is not exist in channels configuration", channel_name.c_str());
+        return false;
+    }
+
+    Channel* activate_channel = nullptr;
+    // get current activated channel
+    if (channels_orderd_map.begin() != channels_orderd_map.end())
```
Comment (webispy) [inline src/core/focus_manager.cc:82]: It seems to be a typo. Please change 'orderd' to 'ordered'.

Comment (JeanTracker) [reply src/core/focus_manager.cc:82]: Thanks. I'll fixed it.

File: include/clientkit/focus_manager_interface.hh (hunk 1 above)
Comment (JeanTracker) [reply include/clientkit/focus_manager_interface.hh:40]: The priority value is freely configurable and has no significant meaning.

File: include/clientkit/nugu_client.hh [hunk 3]
```diff
@@ -121,6 +121,12 @@ public:
      */
     INetworkManager* getNetworkManager();
 
+    /**
+     * @brief Get FocuskManager object
```
Comment (kimhyungrok) [inline include/clientkit/nugu_client.hh:125]: Typo error. Please fix "FocuskManager". The 126 line is same.

File: src/core/focus_manager.hh [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+#include <map>
+#include <memory>
+
+#include "clientkit/focus_manager_interface.hh"
+
+namespace NuguCore {
+
+using namespace NuguClientKit;
+
+class IFocusResourceObserver {
```
Comment (kimhyungrok) [inline src/core/focus_manager.hh:30]: Doesn't this IFocusResourceObserver already declared in focus_manager_interface.hh ?

File: include/clientkit/nugu_client.hh (hunk 3 above)
Comment (JeanTracker) [reply include/clientkit/nugu_client.hh:125]: Thanks. I'll fix it.

**Pull request merged**